File: Problems/Week_2/4.1_도키도키_간식드리미(12789)/임용식/12789.cs
using System;
using System.Collections.Generic;

namespace AlgoCSharp
{
    class Program
    {
        static void Main(string[] args)
        {
            var q               = new Queue<int>();
            var tempLine        = new Stack<int>();
            int n               = Convert.ToInt32(Console.ReadLine());
            int[] personCount   = Array.ConvertAll(Console.ReadLine().Split(), s => Convert.ToInt32(s));

            foreach (var i in personCount)
            {
                q.Enqueue(i);
            }

            int num = 1;

            while (num <= n)
            {
                if (q.Count == 0 && tempLine.Peek() != num)
                {
                    Console.Write("Sad");
                    return;
                }
                if (q.Count != 0 && q.Peek() == num)
                {
                    q.Dequeue();
                    num++;
                }
                else if (tempLine.Count != 0 && tempLine.Peek() == num)
                {
                    tempLine.Pop();
                    num++;
                }
                else
                {
                    while (q.Count > 0 && q.Peek() != num)
                    {
                        tempLine.Push(q.Dequeue());
                    }
                }
            }
            Console.Write("Nice");
        }
    }
}

File: Problems/Week_2/4.1_도키도키_간식드리미(12789)/임용식/12789.cpp
#include <iostream>
#include <queue>
#include <stack>

int main()
{
	std::queue<int> snackLine;
	std::stack<int> tempLine;
	int n;
	std::cin >> n;

	for (int i = 0; i < n; i++)
	{
		int temp;
		std::cin >> temp;
		snackLine.push(temp);
	}

	for (int i = 1; i <= n;)
	{
		if (snackLine.empty() && tempLine.top() != i)
		{
			std::cout << "Sad";
			return 0;
		}

		if (!snackLine.empty() && snackLine.front() == i)
		{
			snackLine.pop();
			i++;
		}
		else if (!tempLine.empty() && tempLine.top() == i)
		{
			tempLine.pop();
			i++;
		}
		else
		{
			while (!snackLine.empty() && snackLine.front() != i)
			{
				tempLine.push(snackLine.front());
				snackLine.pop();
			}
		}
	}
	std::cout << "Nice";

	return 0;
}